server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -481,29 +481,56 @@ private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unas
                         )
                     );
                 }
-                shards.put(shardNumber, indexShardRoutingBuilder.build());
+                shards[shardNumber] = indexShardRoutingBuilder;
             }
             return this;
         }
 
         public Builder addReplica() {
-            for (var shardNumber : shards.keySet()) {
-                ShardId shardId = new ShardId(index, shardNumber);
-                // version 0, will get updated when reroute will happen
-                ShardRouting shard = ShardRouting.newUnassigned(
-                    shardId,
-                    false,
-                    PeerRecoverySource.INSTANCE,
-                    new UnassignedInfo(UnassignedInfo.Reason.REPLICA_ADDED, null)
-                );
-                shards.put(shardNumber, new IndexShardRoutingTable.Builder(shards.get(shard.id())).addShard(shard).build());
+            if (shards == null) {
+                return this;
+            }
+            for (int shardNumber = 0; shardNumber < shards.length; shardNumber++) {
+                Object existing = shards[shardNumber];
+                if (existing == null) {
+                    continue;
+                }
+                if (existing instanceof IndexShardRoutingTable) {
+                    shards[shardNumber] = new IndexShardRoutingTable.Builder((IndexShardRoutingTable) existing).addShard(
+                        newUnassignedReplica(((IndexShardRoutingTable) existing).shardId())
+                    ).build();
+                } else {
+                    assert existing instanceof IndexShardRoutingTable.Builder;
+                    // version 0, will get updated when reroute will happen
+                    ((IndexShardRoutingTable.Builder) existing).addShard(
+                        newUnassignedReplica(((IndexShardRoutingTable.Builder) existing).shardId())
+                    );
+                }
             }
             return this;
         }
 
+        private ShardRouting newUnassignedReplica(ShardId shardId) {
+            return ShardRouting.newUnassigned(
+                shardId,
+                false,
+                PeerRecoverySource.INSTANCE,
+                new UnassignedInfo(UnassignedInfo.Reason.REPLICA_ADDED, null)
+            );
+        }
+
         public Builder removeReplica() {
-            for (var shardId : shards.keySet()) {
-                IndexShardRoutingTable indexShard = shards.get(shardId);
+            if (shards == null) {
+                return this;
+            }
+            for (int shardId = 0; shardId < shards.length; shardId++) {
+                Object found = shards[shardId];
+                if (found == null) {
+                    continue;
+                }
+                final IndexShardRoutingTable indexShard = found instanceof IndexShardRoutingTable
+                    ? (IndexShardRoutingTable) found
+                    : ((IndexShardRoutingTable.Builder) found).build();
```

Comment:
It looks like below this is converted back to the builder. Is it possible to re-use existing builder instead?

Reply:
This complexity is now all gone thanks to moving to an array of builders :)